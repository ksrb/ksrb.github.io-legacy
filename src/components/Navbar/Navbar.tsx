import { Grid } from "@material-ui/core";
import clsx from "clsx";
import React, { FC, useCallback } from "react";
import {
  useScrollProvider,
  useScrollProviderRefCallback,
} from "src/components/ScrollProvider";
import {
  aboutListenerId,
  experiencesListenerId,
  navbarListenerId,
  skillsListenerId,
} from "src/constants";
import Header from "src/components/Header";
import useStyles from "./styles";

const Navbar: FC = () => {
  const scrollProviderRef = useScrollProviderRefCallback(navbarListenerId);

  const {
    [aboutListenerId]: aboutListener,
    [navbarListenerId]: navbarListener,
    [experiencesListenerId]: experiencesListener,
    [skillsListenerId]: skillsListener,
  } = useScrollProvider();

  const classes = useStyles();

  const Items = useCallback(
    () => (
      <>
        <div
          className={clsx(
            classes.item,
            aboutListener.inViewPort && classes.item__active,
          )}
        >
          ABOUT
        </div>
        <div
          className={clsx(
            classes.item,
            skillsListener.inViewPort && classes.item__active,
          )}
        >
          SKILLS
        </div>
        <div
          className={clsx(
            classes.item,
            experiencesListener.inViewPort && classes.item__active,
          )}
        >
          EXPERIENCE
        </div>
      </>
    ),
    [
      aboutListener.inViewPort,
      classes.item,
      classes.item__active,
      experiencesListener.inViewPort,
      skillsListener.inViewPort,
    ],
  );

  return (
    <div className={classes.root} ref={scrollProviderRef}>
      <Grid
        item
        xs={12}
        className={clsx(classes.content, classes.content__centered)}
      >
        <Items />
      </Grid>
      <Grid
        item
        xs={12}
        className={clsx(
          classes.content,
          classes.content__fixed,
          !navbarListener.inViewPort && classes.content__fixedExpanded,
        )}
      >
        <Header classes={{ root: classes.header }} variant="navbar" />
        <Items />
      </Grid>
    </div>
  );
};

export default Navbar;
