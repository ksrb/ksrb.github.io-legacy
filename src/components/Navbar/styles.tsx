import { makeStyles } from "@material-ui/styles";
import defaultTheme, { Theme } from "src/theme";
import { header__navbarHeight } from "src/components/Header/styles";

export const content__fixedExpandedPaddingUsingSpacing = 1;
export const navbarHeight =
  header__navbarHeight +
  defaultTheme.spacing(content__fixedExpandedPaddingUsingSpacing) * 2;

export default makeStyles<Theme>(
  ({
    spacing,
    palette: {
      common: { white },
    },
    marginBottom,
  }) => ({
    content: {
      display: "flex",
      overflow: "hidden",
    },
    content__centered: {
      justifyContent: "center",
      marginBottom: marginBottom,
    },
    content__fixed: {
      display: "flex",
      alignItems: "center",
      position: "fixed",
      width: "100%",
      left: 0,
      top: 0,
      background: white,
      zIndex: 1,
      maxHeight: 0,
      padding: spacing(0, 1, 0, 3),
      transition: "padding-top .3s ease, max-height .3s ease",
    },
    content__fixedExpanded: {
      maxHeight: navbarHeight,
      padding: spacing(
        content__fixedExpandedPaddingUsingSpacing,
        1,
        content__fixedExpandedPaddingUsingSpacing,
        3,
      ),
    },

    item: {
      margin: "0 5px",
    },
    item__active: {
      textDecoration: "underline",
    },

    header: {
      marginRight: spacing(2),
    },
  }),
);
