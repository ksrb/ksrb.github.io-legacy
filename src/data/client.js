import ApolloClient from 'apollo-boost';

const defaults = {
  uiState: {
    __typename: 'UiState',
    selectedItem: 'none',
  },
  experiences: [
    {
      __typename: 'Experience',
      companyName: 'Lantern Credit',
      iconPath: require('data/img/lantern.svg'),
      purpose: 'Startup focused on modernizing the credit scoring industry by creating an application that allows users to modify their credit score in real time.',
      address: {
        __typename: 'Address',
        state: 'CA',
        county: 'Irvine',
      },
      role: 'Full Stack Developer',
      hours: 'Full time',
      startDate: new Date(2015, 6).toString(),
      endDate: new Date(2018, 2).toString(),
      skills: [
        {
          __typename: 'ExperienceSkill',
          name: 'JavaScript (React)',
          type: 'programming',
          utilization: 60,
        },
        {
          __typename: 'ExperienceSkill',
          name: 'Golang',
          type: 'programming',
          utilization: 25,
        },
        {
          __typename: 'ExperienceSkill',
          name: 'Bash',
          type: 'programming',
          utilization: 15,
        },
      ],
      accomplishments: [
        'Used ReactJS, Babel, and Webpack to create front-end application using current and experimental versions of JavaScript.',
        'Maintained front end build and deployment pipeline and incorporated new technologies to streamline development such as SASS and CSS modules.',
        'Created several microservices using Golang to retrieve data from credit agencies such as Transunion and Equifax. ',
        'Built and deployed microservices using Docker, took part in the creation of a 3-tier architecture creating a system that was both secure and scalable.',
      ],
    },
    {
      __typename: 'Experience',
      companyName: 'Table Design Art',
      iconPath: require('data/img/table design art.png'),
      purpose: 'Startup company specializing in creating decorative table banners for special occasion.',
      address: {
        __typename: 'Address',
        state: 'CA',
        county: 'Irvine',
      },
      role: 'Web Developer',
      hours: 'Part time',
      startDate: new Date(2015, 3).toString(),
      endDate: new Date(2015, 6).toString(),
      skills: [
        {
          __typename: 'ExperienceSkill',
          name: 'Liquid',
          type: 'visual',
          utilization: 40,
        },
        {
          __typename: 'ExperienceSkill',
          name: 'css',
          type: 'visual',
          utilization: 30,
        },
        {
          __typename: 'ExperienceSkill',
          name: 'JavaScript',
          type: 'programming',
          utilization: 30,
        },
      ],
      accomplishments: [
        'Used Shopify to rapidly create webstore, used the Liquid templating language to customize appearance.',
        'Simplified shipping process by integrating webstore with shipping providers such as FedEx and UPS.',
        'Helping manage social media campaign on Facebook, Pinterest, and Twitter.',
      ],
    },
    {
      __typename: 'Experience',
      companyName: 'Niksun',
      iconPath: require('data/img/niksun.png'),
      purpose: 'Company focused on providing network analysis software.',
      address: {
        __typename: 'Address',
        state: 'CA',
        county: 'Irvine',
      },
      role: 'Web Developer',
      hours: 'Full time',
      startDate: new Date(2013, 9).toString(),
      endDate: new Date(2015, 2).toString(),
      skills: [
        {
          __typename: 'ExperienceSkill',
          name: 'GWT',
          type: 'programming',
          utilization: 70,
        },
        {
          __typename: 'ExperienceSkill',
          name: 'JavaScript',
          type: 'programming',
          utilization: 20,
        },
        {
          __typename: 'ExperienceSkill',
          name: 'css',
          type: 'visual',
          utilization: 10,
        },
      ],
      accomplishments: [
        'Created a real time reporting web application for Niksun\'s file  analysis service, used by several clients including Wells Fargo.',
        'Rapidly designed and wireframed several prototypes in an effort to modernize Niksun\'s client facing applications.',
        'Created a workspace setup and build script using Gradle, significantly reducing onboarding time of new developers.',
      ],
    }
  ],
  skills: [
    // programming
    {
      __typename: 'Skills',
      name: 'JavaScript',
      iconPath: require('data/img/javascript.svg'),
      level: 4.5,
      comment: 'Almost expert',
      type: 'programming',
      link: 'https://en.wikipedia.org/wiki/JavaScript'
    },
    {
      __typename: 'Skills',
      name: 'ReactJS',
      iconPath: require('data/img/reactjs.svg'),
      level: 3.5,
      comment: 'Haven\'t taught yet',
      type: 'programming',
      link: 'https://en.wikipedia.org/wiki/React_(JavaScript_library)'
    },
    {
      __typename: 'Skills',
      name: 'Golang',
      iconPath: require('data/img/golang.svg'),
      level: 3,
      comment: 'Proficient',
      type: 'programming',
      link: 'https://en.wikipedia.org/wiki/Go_(programming_language)',
    },
    {
      __typename: 'Skills',
      name: 'Java',
      iconPath: require('data/img/java.svg'),
      level: 3,
      comment: 'Proficient',
      type: 'programming',
      link: 'https://en.wikipedia.org/wiki/Java_(programming_language)',
    },
    {
      __typename: 'Skills',
      name: 'Bash',
      iconPath: require('data/img/bash.svg'),
      level: 3,
      comment: 'Proficient',
      type: 'programming',
      link: 'https://en.wikipedia.org/wiki/Bash_(Unix_shell)'
    },
    // visual
    {
      __typename: 'Skills',
      name: 'Sass',
      iconPath: require('data/img/sass.svg'),
      level: 3.5,
      comment: 'Proficient',
      type: 'visual',
      link: 'https://en.wikipedia.org/wiki/Sass_(stylesheet_language)'
    },
    {
      __typename: 'Skills',
      name: 'Photoshop',
      iconPath: require('data/img/photoshop.svg'),
      level: 3.5,
      comment: 'Proficient',
      type: 'visual',
      link: 'https://en.wikipedia.org/wiki/Adobe_Photoshop'
    },
    {
      __typename: 'Skills',
      name: 'Illustrator',
      iconPath: require('data/img/illustrator.svg'),
      level: 2.5,
      comment: 'Proficient',
      type: 'visual',
      link: 'https://en.wikipedia.org/wiki/Adobe_Illustrator',
    },
    // tool
    {
      __typename: 'Skills',
      name: 'IntelliJ',
      iconPath: require('data/img/intellij.svg'),
      level: 3,
      comment: 'Proficient',
      type: 'tool',
      link: 'https://en.wikipedia.org/wiki/IntelliJ_IDEA',
    },
    {
      __typename: 'Skills',
      name: 'vim',
      iconPath: require('data/img/neovim.png'),
      level: 3,
      comment: 'Proficient',
      type: 'tool',
      link: 'https://en.wikipedia.org/wiki/Vim_(text_editor)#Neovim',
    },
    {
      __typename: 'Skills',
      name: 'git',
      iconPath: require('data/img/git.svg'),
      level: 4,
      comment: 'Taught at Niksun',
      type: 'tool',
      link: 'https://en.wikipedia.org/wiki/Git'
    },

  ],
};

const Mutation = {
  // https://www.apollographql.com/docs/graphql-tools/resolvers
  changeSelectedNavbarItem(obj, args, context) {
    const { cache } = context;
    cache.writeData({
      data: {
        uiState: {
          __typename: 'UiState',
          selectedItem: args.item
        }
      }
    });
    return null;
  },
};

const Query = {
  uiState: (obj, args, { cache }) => {
    return cache.uiState;
  },
  experiences: (obj, args, { cache }) => {
    return cache.experiences;
  },
  skills: (obj, args, { cache }) => {
    return cache.skills;
  },
};

const typeDefs = `
  type UIState {
    selectedItem: Int
    searchedSkill: String
  }

  type Experience {
    companyName: String
    iconPath: String
    purpose: String
    address: Address
    role: String
    hours: String
    startDate: String
    endDate: String
    skills: [ExperienceSkill]
    accomplishments: [String]
  }

  type Address {
    state: String
    county: String
  }

  type ExperienceSkill {
    name: String
    utilization: Int
  }
  
  type Skill {
    name: String
    iconPath: String
    level: Float
    comment: string
    type: String
  }

  type Query {
    uiState: UIState
    experiences: [Experience]
    skills: [Skill]
  }

  type Mutation {
    changeSelectedNavbarItem(item: String!): String
  }
`;

const client = new ApolloClient({
  clientState: {
    defaults,
    resolvers: {
      Mutation,
      Query,
    },
    typeDefs,
  },
});

export default client;
