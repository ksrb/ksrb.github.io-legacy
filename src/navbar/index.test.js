import React from 'react';
import ReactDOM from 'react-dom';
import { ApolloProvider } from "react-apollo";
import { Mutation } from 'react-apollo';
import { gql } from 'apollo-boost';

import renderer from 'react-test-renderer';

import Navbar from './index';
import client from 'data/client';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <ApolloProvider client={client}>
      <Navbar />
    </ApolloProvider>,
    div);
  ReactDOM.unmountComponentAtNode(div);
});

const MUTATE_SELECTED_NAVBAR_ITEM = gql`
  mutation changeSelectedNavbarItem($item: String!) {
    changeSelectedNavbarItem(item: $item) @client {
      uiState {
        selectedItem
      }
    }
  }
`;

it('changes navbar item style when item clicked', (done) => {
  const component = renderer.create(
    <ApolloProvider client={client}>
      <Navbar />
      <Mutation mutation={MUTATE_SELECTED_NAVBAR_ITEM}>
        {(mutate) => {
          mutate({ variables: { item: 'ABOUT' } }).then(() => {
            done();
          });
          return (
            <div />
          );
        }}
      </Mutation>
    </ApolloProvider>);

  let tree = component.toJSON();

  let aboutItem = tree.children[1].children[0];
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
