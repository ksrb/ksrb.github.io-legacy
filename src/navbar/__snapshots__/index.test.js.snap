// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes navbar item style when item clicked 1`] = `
<div
  className="root"
>
  <div
    className="logo"
  >
    <div
      className="logo_icon"
    />
    <div
      className="logo_text"
    >
      K
      <span
        className="logo_text_small"
      >
        EVIN
      </span>
      Â  S
      <span
        className="logo_text_small"
      >
        UEN
      </span>
    </div>
  </div>
  <div
    className="navbar"
  >
    <div
      className="navbar_item "
      onClick={[Function]}
    >
      ABOUT
    </div>
    <div
      className="navbar_item "
      onClick={[Function]}
    >
      SKILLS
    </div>
    <div
      className="navbar_item "
      onClick={[Function]}
    >
      EXPERIENCE
    </div>
    <div
      className="navbar_item "
      onClick={[Function]}
    >
      EDUCATION
    </div>
  </div>
</div>
`;
